crates/release_plz_core/src/git/backend.rs

```diff
@@ -540,49 +563,163 @@ impl GitClient {
+    }
+
+    /// Add all labels to PR
+    async fn post_github_labels(&self, labels: &[String], pr_number: u64) -> anyhow::Result<()> {
+        self.client
+            .post(self.pr_labels_url(pr_number))
+            .json(&json!({
+                "labels": labels
+            }))
+            .send()
+            .await?
+            .successful_status()
+            .await?;
+
+        Ok(())
+    }
+
+    /// Add all labels to PR
+    async fn post_gitlab_labels(&self, labels: &[String], pr_number: u64) -> anyhow::Result<()> {
+        self.client
+            .put(format!("{}/{}", self.pulls_url(), pr_number))
+            .json(&json!({
+                "add_labels": labels.iter().join(",")
+            }))
+            .send()
+            .await?
+            .successful_status()
+            .await?;
 
-                Ok(())
+        Ok(())
+    }
+
+    /// Add all labels to PR
+    async fn post_gitea_labels(&self, label_ids: &[u64], pr_number: u64) -> anyhow::Result<()> {
+        self.client
+            .post(self.pr_labels_url(pr_number))
+            .json(&json!({ "labels": label_ids }))
+            .send()
+            .await?
+            .successful_status()
+            .await?;
+        Ok(())
+    }
+
+    async fn get_pr_info_and_categorize_labels(
+        &self,
+        pr_number: u64,
+        labels: &[String],
+    ) -> anyhow::Result<(Vec<String>, Vec<u64>)> {
+        let current_pr_info = self
+            .get_pr_info(pr_number)
+            .await
+            .context("failed to get pr info")?;
+        let existing_labels = current_pr_info.labels;
+
+        let label_map: HashMap<String, Option<u64>> = existing_labels
+            .iter()
+            .map(|l| (l.name.clone(), l.id))
+            .collect();
+
+        let mut labels_to_create = Vec::new();
+        let mut label_ids = Vec::new();
+
+        for label in labels {
+            match label_map.get(label) {
+                Some(id) => label_ids
+                    .push(id.with_context(|| format!("failed to extract id from label {label}"))?),
+                None => labels_to_create.push(label.to_owned()),
             }
-            BackendType::Gitea => {
-                warn!("PR labels are only supported on Github and Gitlab");
-                Ok(())
+        }
+
+        Ok((labels_to_create, label_ids))
+    }
+
+    async fn create_gitea_labels(&self, labels_to_create: &[String]) -> anyhow::Result<Vec<u64>> {
+        let mut label_ids = Vec::new();
+
+        for label in labels_to_create {
+            debug!("Backend Gitea creating label: {label}");
+            let res = self
+                .client
+                .post(format!("{}/labels", self.repo_url()))
+                .json(&json!({
+                    "name": label.trim(),
+                    "color": "#FFFFFF"
+                }))
+                .send()
+                .await?
+                .successful_status()
```

Comment:
the successful_status function returns an error in case of NOT_FOUND and UNPROCESSABLE_ENTITY so those match arms are unreachable because in the next line you use ?

I think you should match on the error, and then match on the status.
Then when you match on the error, add context instead of creating a new error, so still show the error returned from Gitea to the user, as it might contain useful info.

Reply:
Nice. Thank you for pointing this out